src/sentry/sentry_metrics/consumers/indexer/routing_producer.py

```diff
@@ -0,0 +1,166 @@
+        self.__closed = False
+        self.__offsets_to_be_committed: MutableMapping[Partition, Position] = {}
+        self.__queue: Deque[Tuple[Partition, Position, Future[Message[KafkaPayload]]]] = deque()
+
+    def poll(self) -> None:
+        """
+        Periodically check which messages have been produced successfully and
+        call the commit function for the offsets that have been processed.
+        In order to commit offsets we need strict ordering, hence we stop
+        processing the queue of pending futures as soon as we encounter one
+        which is not yet completed.
+        """
+        while self.__queue:
+            partition, position, future = self.__queue[0]
+
+            if not future.done():
+                break
+
+            future.result()
+
+            self.__queue.popleft()
+            self.__offsets_to_be_committed[partition] = position
+
+        self.__commit_function(self.__offsets_to_be_committed)
+
+    def submit(self, message: Message[RoutingPayload]) -> None:
+        """
+        This is where the actual routing happens. Each message is passed to the
+        message router to get the producer and topic to which the message should
+        be sent. The message is then sent to the producer and the future is
+        stored in the queue.
+        """
+        assert not self.__closed
+        producer, topic = self.__message_router.get_route_for_message(message)
+        output_message = Message(
+            partition=message.partition,
+            offset=message.offset,
+            timestamp=message.timestamp,
+            payload=message.payload.routing_message.value,
+        )
+
+        self.__queue.append(
+            (
+                output_message.partition,
+                output_message.position_to_commit,
+                producer.produce(destination=topic, payload=output_message.payload),
+            )
+        )
+
+    def terminate(self) -> None:
+        self.__closed = True
+
+    def close(self) -> None:
+        self.__closed = True
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        """
+        In addition to flushing the queue, this method also calls the
+        shutdown of the router.
+        """
+        start = time.time()
+
+        # Commit all previously staged offsets
+        self.__commit_function({}, force=True)
+
+        while self.__queue:
+            remaining = timeout - (time.time() - start) if timeout is not None else None
+            if remaining is not None and remaining <= 0:
+                logger.warning(f"Timed out with {len(self.__queue)} futures in queue")
+                break
+
+            partition, position, future = self.__queue.popleft()
+            future.result(remaining)
+            offset = {partition: position}
+
+            logger.info("Committing offset: %r", offset)
+            self.__commit_function(offset, force=True)
+
+        self.__message_router.shutdown()
```

Comment:
I think you should remove this line. You can call it from the consumer later when this is integrated. Since this step is getting passed the message router it is not its job to attempt to shut it down over and over on every balance. It's also ignoring the timeout now if you call it from here.

Reply:
I have removed it for now. Once I integrate with the consumer builder, we can see where to best put it.